docs/configuration.md
# Configuration

This describes `sb`'s main configuration options.

## Commands

This block only describes the command (as in [cmd/*.go](../cmd/)) that is executed when connecting to a host.

```yaml
commands:
  ssh_command: ttyrec
```

Right now, the only valid option is `ttyrec`: it will connect you to the distant host via `ssh` while recording 
the session with `ttyrec`.

## General

```yaml
general:
  binary_path: /opt/sb/sb
  sb_user: sb
  sb_user_home: /home/sb
  hostname: sb.YOUR_DOMAIN.com
  name: sb-us
  location: "us"
  ssh_port: 22
  mosh_ports_range: 40000:49999
  env_vars_to_forward: ["USER"]
  encryption-key: changemechangemechangemechangeme
```

- `binary_path` (string): the path where `sb`'s binary is on the bastion server
- `sb_user` (string): the main `sb` system user
- `sb_user_home` (string): the home of the main `sb` user (where global logs are stored)
- `hostname` (string): the hostname of this instance; it will be displayed in the information
- `name` (string): the name of this instance; it will be displayed in the information
- `location` (string): the location of this instance; it will be displayed in the information
- `ssh_port` (int): the port SSH listens on for this instance
- `mosh_port_range` (string): the UDP range ports that [Mosh](https://github.com/mobile-shell/mosh) can use
- `env_vars_to_forward` ([]string): the environment variables that `sb` will forward to a distant host
- `encryption-key` (string): the encryption key for replication, TTYRecs offloading and backups; 
  it must be either 16, 24 or 32 characters

## Replication

To learn about replication and high availability, please refer 
to [the dedicated documentation](./high-availability.md).

```yaml
replication:
  enabled: false
  queue:
    type: googlepubsub
    googlepubsub:
      project: demo-gcp-project
      topic: sb
```

- `enabled` (bool): whether or not replication is enabled
- `queue`:
  - `type` (string): the type of queue to use; as of today, only `googlepubsub`
  - `googlepubsub`:
    - `project` (string): The GCP project that hosts the Google PubSub queue
    - `topic` (string): The queue's topic

To avoid multiple instances pulling from the same Google PubSub subscription, each `sb` instance registers 
to a `subscription` built on the following model: `[topic-name]-[linux-hostname]`.

If the subscription doesn't exist, it is set to be created automatically.

## TTYRecs offloading

To learn about TTYRecs offloading and high availability, please refer 
to [the dedicated documentation](./high-availability.md#ttyrecs-offloading).

```yaml
ttyrecsoffloading:
  enabled: false
  storage:
    type: gcs
    gcs:
      bucket: sb-bucket
      objects-base-path: ttyrecs
    s3:
      region: us-east-2
      bucket: sb-bucket
      keys-base-path: ttyrecs
      aws-access-key: xxx
      aws-secret-key: xxx
      aws-session-token: xxx
```

- `enabled` (bool): whether or not TTYRecs offloading is enabled
- `storage`:
  - `type` (string): the type of remote storage to use; as of today only `gcs` or `s3`
  - `gcs`:
    - `bucket` (string): the Google Cloud Storage bucket to use
    - `objects-base-path` (string): the objects' prefix in the bucket
  - `s3`:
    - `region` (string): the region the Amazon S3 bucket sits
    - `bucket` (string): the Amazon S3 bucket to use
    - `keys-base-path` (string): the objects' prefix in the bucket
    - `aws-access-key` (string): optional AWS access key; if not specified, taken from the environment
    - `aws-secret-key` (string): optional AWS secret key; if not specified, taken from the environment
    - `aws-session-token` (string): optional AWS session token to use; if not specified, taken from the environment


internal/config/config.go
package config

import (
	"fmt"
	"os"

	"github.com/inpher/sb/internal/types"
	"github.com/spf13/viper"
)

var (
	VERSION string
	COMMIT  string
)

// Initialize initializes the viper config and sets default values
func init() {

	viper.SetConfigName("sb")
	viper.SetConfigType("yaml")
	viper.AddConfigPath(".")
	viper.AddConfigPath("/etc/sb")

	err := viper.ReadInConfig()
	if err != nil {

		if _, ok := err.(viper.ConfigFileNotFoundError); ok {

			// General instance configuration
			viper.SetDefault("general.name", "sb")
			viper.SetDefault("general.location", "earth")
			viper.SetDefault("general.hostname", "sb.domain.tld")
			viper.SetDefault("general.binary_path", "/opt/sb/sb")
			viper.SetDefault("general.ssh_port", "22")
			viper.SetDefault("general.mosh_ports_range", "40000:49999")
			viper.SetDefault("general.env_vars_to_forward", []string{"USER"})
			viper.SetDefault("general.sb_user", "sb")
			viper.SetDefault("general.sb_user_home", "/home/sb")
			viper.SetDefault("general.encryption-key", "changemechangemechangemechangeme")

			// Commands configuration
			viper.SetDefault("commands.ssh_command", "ttyrec")

			// Replication configuration
			viper.SetDefault("replication.enabled", false)
			viper.SetDefault("replication.queue.type", "")
			viper.SetDefault("replication.queue.googlepubsub.project", "")
			viper.SetDefault("replication.queue.googlepubsub.topic", "")

			// TTYrecs offloading configuration
			viper.SetDefault("ttyrecsoffloading.enabled", false)
			viper.SetDefault("ttyrecsoffloading.storage.type", "")
			viper.SetDefault("ttyrecsoffloading.storage.gcs.bucket", "")
			viper.SetDefault("ttyrecsoffloading.storage.gcs.objects-base-path", "")
			viper.SetDefault("ttyrecsoffloading.storage.gcs.endpoint-url", "")
			viper.SetDefault("ttyrecsoffloading.storage.s3.bucket", "")
			viper.SetDefault("ttyrecsoffloading.storage.s3.keys-base-path", "")

		} else {

			os.Exit(-1)

		}
	}
}

// GetSBName returns sb's name (AKA the alias to set in user's path)
func GetSBName() string {
	return viper.GetString("general.name")
}

// GetSBLocation returns sb's location (AKA the specific instance in a replicated set)
func GetSBLocation() string {
	return viper.GetString("general.location")
}

// GetSBHostname returns sb's hostname
func GetSBHostname() string {
	return viper.GetString("general.hostname")
}

// GetSSHCommand returns the SSH command to launch when user wants to access a host
func GetSSHCommand() string {
	return viper.GetString("commands.ssh_command")
}

// GetMOSHPortsRange returns the MOSH server ports range
func GetMOSHPortsRange() string {
	return viper.GetString("general.mosh_ports_range")
}

// GetSSHPort returns the SSH server port
func GetSSHPort() string {
	return viper.GetString("general.ssh_port")
}

// GetEnvironmentVarsToForward returns the list of environment variables to forward to distant hosts
func GetEnvironmentVarsToForward() []string {
	return viper.GetStringSlice("general.env_vars_to_forward")
}

// GetBinaryPath returns the path of the sb binary
func GetBinaryPath() string {
	return viper.GetString("general.binary_path")
}

// GetSBUsername returns the global sb user name
func GetSBUsername() string {
	return viper.GetString("general.sb_user")
}

// GetSBUserHome returns the global sb user home
func GetSBUserHome() string {
	return viper.GetString("general.sb_user_home")
}

// GetGlobalDatabasePath returns the global database path
func GetGlobalDatabasePath() string {
	return fmt.Sprintf("%s/logs.db", GetSBUserHome())
}

// GetReplicationDatabasePath returns the global database path
func GetReplicationDatabasePath() string {
	return fmt.Sprintf("%s/replication.db", GetSBUserHome())
}

// GetEncryptionKey returns the symmetric encryption key for backup, replication and ttyrecs offloading
func GetEncryptionKey() string {
	return viper.GetString("general.encryption-key")
}

func GetReplicationEnabled() bool {
	return viper.GetBool("replication.enabled")
}

func GetReplicationQueueConfig() *types.ReplicationQueueConfig {
	return &types.ReplicationQueueConfig{
		Enabled:      viper.GetBool("replication.enabled"),
		QueueType:    viper.GetString("replication.queue.type"),
		QueueOptions: viper.Sub(fmt.Sprintf("replication.queue.%s", viper.GetString("replication.queue.type"))),
	}
}

func GetTTYRecsOffloadingConfig() *types.TTYRecsOffloadingConfig {
	return &types.TTYRecsOffloadingConfig{
		Enabled:        viper.GetBool("ttyrecsoffloading.enabled"),
		StorageType:    viper.GetString("ttyrecsoffloading.storage.type"),
		StorageOptions: viper.Sub(fmt.Sprintf("ttyrecsoffloading.storage.%s", viper.GetString("ttyrecsoffloading.storage.type"))),
	}
}
